audio/iir_filter_node.rs

```diff
@@ -0,0 +1,177 @@
+use block::Chunk;
+use node::{AudioNodeEngine, AudioNodeType, BlockInfo, ChannelInfo};
+use std::sync::Arc;
+use std::collections::VecDeque;
+
+use num_complex::Complex64;
+
+const MAX_COEFFS: usize = 20;
+
+#[derive(Debug)]
+pub struct IIRFilterNodeOptions {
+    pub feedforward: Vec<f64>,
+    pub feedback: Vec<f64>,
+}
+
+#[derive(Clone)]
+struct IIRFilter {
+    feedforward: Arc<Vec<f64>>,
+    feedback: Arc<Vec<f64>>,
+    inputs: VecDeque<f64>,
+    outputs: VecDeque<f64>,
+}
+
+impl IIRFilter {
+    fn new(feedforward: Arc<Vec<f64>>, feedback: Arc<Vec<f64>>) -> Self {
+        Self {
+            feedforward,
+            feedback,
+            inputs: VecDeque::with_capacity(MAX_COEFFS),
+            outputs: VecDeque::with_capacity(MAX_COEFFS),
+        }
+    }
+
+    fn calculate_output(&mut self, input: f32) -> f32 {
+        self.inputs.push_front(input as f64);
+
+        if self.inputs.len() > MAX_COEFFS {
+            self.inputs.pop_back();
+        }
+
+        let inputs_sum = self
+            .feedforward
+            .iter()
+            .zip(self.inputs.iter())
+            .fold(0.0, |acc, (c, v)| acc + c * v);
+
+        let outputs_sum = self
+            .feedback
+            .iter()
+            .skip(1)
+            .zip(self.outputs.iter())
+            .fold(0.0, |acc, (c, v)| acc + c * v);
+
+        let output = (inputs_sum - outputs_sum) / self.feedback[0];
+
+        if output.is_nan() {
+            // Per spec:
+            // Note: The UA may produce a warning to notify the user that NaN values have occurred in the filter state.
+            // This is usually indicative of an unstable filter.
+            //
+            // But idk how to produce warnings
+            panic!("NaN in IIRFilter state");
```

Comment:
A panic here may be too much. Let's `warn!` instead.

Unfortunately, AFAICT we have no way to progress this warning up to the DOM code at the moment.